proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3035,256 @@ lemma cancelIPC_ccorres1:
+  subgoal by (auto simp: invs'_implies valid_tcb_state'_def BlockedOnReceive_replyObject_no_0)
     apply (rule conjI)
-     subgoal by (auto simp: projectKOs obj_at'_def pred_tcb_at'_def split: thread_state.splits)[1]
-    apply (clarsimp)
+     apply (clarsimp simp: inQ_def)
+    apply clarsimp
     apply (rule conjI)
-     subgoal by (auto simp: obj_at'_def projectKOs pred_tcb_at'_def invs'_def valid_state'_def
-                     isTS_defs cte_wp_at_ctes_of
-                     cthread_state_relation_def sch_act_wf_weak valid_ep'_def
-                     split: thread_state.splits)
+     apply (clarsimp simp: invs_valid_objs' BlockedOnSend_ep_at')
     apply clarsimp
-    apply (frule (2) ep_blocked_in_queueD_recv)
+    apply (rule conjI)
+     apply (clarsimp simp: sym_refs_asrt_def invs'_implies)
+  subgoal by (auto simp: obj_at'_def pred_tcb_at'_def valid_ep'_def isTS_defs isSendEP_def
+                  split: thread_state.splits endpoint.splits)
+    apply (clarsimp simp: sym_refs_asrt_def invs'_implies valid_objs'_valid_tcbs'
+                          sym_ref_BlockedOnReceive_replyObject_linked)
+    apply (frule (3) ep_blocked_in_queueD_send)
     apply (frule (1) ko_at_valid_ep'[OF _ invs_valid_objs'])
-    subgoal by (auto simp: obj_at'_def projectKOs pred_tcb_at'_def invs'_def valid_state'_def
-                         isTS_defs cte_wp_at_ctes_of isRecvEP_def
-                         cthread_state_relation_def sch_act_wf_weak valid_ep'_def
-                    split: thread_state.splits endpoint.splits)
-   apply (rule conjI)
-    apply (clarsimp simp: inQ_def)
-   apply clarsimp
-   apply (rule conjI)
-    subgoal by (auto simp: obj_at'_def projectKOs pred_tcb_at'_def invs'_def valid_state'_def
-                         isTS_defs cte_wp_at_ctes_of
-                         cthread_state_relation_def sch_act_wf_weak valid_ep'_def
-                    split: thread_state.splits)
-   apply clarsimp
-   apply (rule conjI)
-    subgoal by (auto simp: obj_at'_def projectKOs pred_tcb_at'_def invs'_def valid_state'_def
-                         isTS_defs cte_wp_at_ctes_of
-                         cthread_state_relation_def sch_act_wf_weak valid_ep'_def
-                    split: thread_state.splits)
-   apply clarsimp
-   apply (frule (2) ep_blocked_in_queueD_send)
-   apply (frule (1) ko_at_valid_ep'[OF _ invs_valid_objs'])
-   subgoal by (auto simp: obj_at'_def projectKOs pred_tcb_at'_def invs'_def valid_state'_def
-                        isTS_defs cte_wp_at_ctes_of isSendEP_def
-                        cthread_state_relation_def sch_act_wf_weak valid_ep'_def
-                   split: thread_state.splits endpoint.splits)[1]
-  apply (auto simp: isTS_defs cthread_state_relation_def typ_heap_simps weak_sch_act_wf_def)
-  apply (case_tac ts,
-           auto simp: isTS_defs cthread_state_relation_def typ_heap_simps)
-  done *)
+  subgoal by (auto simp: obj_at'_def pred_tcb_at'_def valid_ep'_def isTS_defs isSendEP_def
```

Comment:
ok, and final nitpick, which is the most annoying to fix up: the current auto-indenter shoves `subgoal` was off to the left, regardless of how deep you were in the proof, which requires manually indenting the `subgoal proof`, i.e. it should be:

```
    apply (rule conjI)
     apply (clarsimp simp: sym_refs_asrt_def invs'_implies)
     subgoal by (auto simp: obj_at'_def pred_tcb_at'_def valid_ep'_def isTS_defs isSendEP_def
                     split: thread_state.splits endpoint.splits)
    apply (clarsimp simp: sym_refs_asrt_def invs'_implies valid_objs'_valid_tcbs'
                          sym_ref_BlockedOnReceive_replyObject_linked)
```
because the subgoal discharges one subgoal. This is in the style guide, but the tool works against us here (there is of course a thread on the Isabelle mailing list about this that went nowhere).

Reply:
Done.